ArduCopter/mode_systemid.cpp

```diff
@@ -0,0 +1,319 @@
+    // @User: Standard
+    // @Values: 0:None, 1:Input Roll Angle, 2:Input Pitch Angle, 3:Input Yaw Angle, 4:Recovery Roll Angle, 5:Recovery Pitch Angle, 6:Recovery Yaw Angle, 7:Disturbance Roll, 8:Disturbance Pitch, 9:Disturbance Yaw, 10:Rate Roll, 11:Rate Pitch, 12:Rate Yaw, 13:Mixer Roll, 14:Mixer Pitch, 15:Mixer Yaw, 16:Mixer Thrust
+    AP_GROUPINFO("_AXIS", 1, ModeSystemId, systemID_axis, 0),
+
+    // @Param: _MAG
+    // @DisplayName: Chirp Magnitude
+    // @Description: Magnitude of sweep in deg, deg/s and 0-1 for mixer outputs.
+    // @User: Standard
+    AP_GROUPINFO("_MAG", 2, ModeSystemId, magnitude, 15),
+
+    // @Param: _F_START_HZ
+    // @DisplayName: Start Frequency
+    // @Description: Frequency at the start of the sweep
+    // @Range: 0.01 100
+    // @Units: Hz
+    // @User: Standard
+    AP_GROUPINFO("_F_START_HZ", 3, ModeSystemId, fStart, 0.5f),
+
+    // @Param: _F_STOP_HZ
+    // @DisplayName: Stop Frequency
+    // @Description: Frequency at the end of the sweep
+    // @Range: 0.01 100
+    // @Units: Hz
+    // @User: Standard
+    AP_GROUPINFO("_F_STOP_HZ", 4, ModeSystemId, fStop, 40),
+
+    // @Param: _T_FADE_IN
+    // @DisplayName: Fade in time
+    // @Description: Time to reach maximum amplitude of sweep
+    // @Range: 0 20
+    // @Units: s
+    // @User: Standard
+    AP_GROUPINFO("_T_FADE_IN", 5, ModeSystemId, tFadeIn, 15),
+
+    // @Param: _T_REC
+    // @DisplayName: Total Sweep length
+    // @Description: Time taken to complete the sweep
+    // @Range: 0 255
+    // @Units: s
+    // @User: Standard
+    AP_GROUPINFO("_T_REC", 6, ModeSystemId, tRec, 70),
+
+    // @Param: _T_FADE_OUT
+    // @DisplayName: Fade out time
+    // @Description: Time to reach zero amplitude at the end of the sweep
+    // @Range: 0 5
+    // @Units: s
+    // @User: Standard
+    AP_GROUPINFO("_T_FADE_OUT", 7, ModeSystemId, tFadeOut, 2),
+
+    AP_GROUPEND
+};
+
+#define SYSTEM_ID_DELAY     1.0f      // speed below which it is always safe to switch to loiter
+
+// systemId_init - initialise systemId controller
+bool ModeSystemId::init(bool ignore_checks)
+{
+    // if landed and the mode we're switching from does not have manual throttle and the throttle stick is too high
+    if (motors->armed() && copter.ap.land_complete && !copter.flightmode->has_manual_throttle()) {
+        return false;
+    }
+
+    orig_bf_feedforward = attitude_control->get_bf_feedforward();
+    waveformTime = 0.0f;
+    magnitude_scale = 0.0f;
+    tConst = 2.0f / fStart; // Two full cycles at the starting frequency
+    systemIDState = SystemID_Testing;
+
+    gcs().send_text(MAV_SEVERITY_INFO, "SystemID Starting: axis=%d", (unsigned)systemID_axis);
+
+    AP::logger().Write(
+        "SID",
+        "TimeUS,Ax,FSt,FSp,TFin,TC,TR,TFout",
+        "s--ssssss",
+        "F--------",
+        "Qdfffffff",
+        AP_HAL::micros64(),
+        (uint8_t)systemID_axis,
+        (double)magnitude,
```

Comment:
why are you casting these to double?  why is that necessary?

Reply:
To be honest I am not sure. This seems to be the way it is done even when they are floats.